# IFT 799 - Science de données
## TP 1 : Compréhension et visualisation des données

Travail réalisé par :
- **ASSABAR Taoufik** --- *asst1001*
- **KHALOUI Oussama** --- *khao1201*

In [559]:
import pandas as pd
import numpy as np

In [560]:
# Read CSV data from files
df1 = pd.read_csv('labels.csv', index_col=0)
df2 = pd.read_csv('data.csv', index_col=0)


# Merge DataFrames based on index (sample names)
merged_df = pd.merge(df1, df2, left_index=True, right_index=True)

# Print the merged DataFrame
merged_df

Class  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \
sample_0    PRAD     0.0  2.017209  3.265527  5.478487  10.431999     0.0   
sample_1    LUAD     0.0  0.592732  1.588421  7.586157   9.623011     0.0   
sample_2    PRAD     0.0  3.511759  4.327199  6.881787   9.870730     0.0   
sample_3    PRAD     0.0  3.663618  4.507649  6.659068  10.196184     0.0   
sample_4    BRCA     0.0  2.655741  2.821547  6.539454   9.738265     0.0   
...          ...     ...       ...       ...       ...        ...     ...   
sample_796  BRCA     0.0  1.865642  2.718197  7.350099  10.006003     0.0   
sample_797  LUAD     0.0  3.942955  4.453807  6.346597  10.056868     0.0   
sample_798  COAD     0.0  3.249582  3.707492  8.185901   9.504082     0.0   
sample_799  PRAD     0.0  2.590339  2.787976  7.318624   9.987136     0.0   
sample_800  PRAD     0.0  2.325242  3.805932  6.530246   9.560367     0.0   

              gene_6    gene_7  gene_8  ...  gene_20521  gene_20522  \
sample_0    7.175175  0.591871     0.0  ...    4.926711    8.210257   
sample_1    6.816049  0.000000     0.0  ...    4.593372    7.323865   
sample_2    6.972130  0.452595     0.0  ...    5.125213    8.127123   
sample_3    7.843375  0.434882     0.0  ...    6.076566    8.792959   
sample_4    6.566967  0.360982     0.0  ...    5.996032    8.891425   
...              ...       ...     ...  ...         ...         ...   
sample_796  6.764792  0.496922     0.0  ...    6.088133    9.118313   
sample_797  7.320331  0.000000     0.0  ...    6.371876    9.623335   
sample_798  7.536589  1.811101     0.0  ...    5.719386    8.610704   
sample_799  9.213464  0.000000     0.0  ...    5.785237    8.605387   
sample_800  7.957027  0.000000     0.0  ...    6.403075    8.594354   

            gene_20523  gene_20524  gene_20525  gene_20526  gene_20527  \
sample_0      9.723516    7.220030    9.119813   12.003135    9.650743   
sample_1      9.740931    6.256586    8.381612   12.674552   10.517059   
sample_2     10.908640    5.401607    9.911597    9.045255    9.788359   
sample_3     10.141520    8.942805    9.601208   11.392682    9.694814   
sample_4     10.373790    7.181162    9.846910   11.922439    9.217749   
...                ...         ...         ...         ...         ...   
sample_796   10.004852    4.484415    9.614701   12.031267    9.813063   
sample_797    9.823921    6.555327    9.064002   11.633422   10.317266   
sample_798   10.485517    3.589763    9.350636   12.180944   10.681194   
sample_799   11.004677    4.745888    9.626383   11.198279   10.335513   
sample_800   10.243079    9.139459   10.102934   11.641081   10.607358   

            gene_20528  gene_20529  gene_20530  
sample_0      8.921326    5.286759    0.000000  
sample_1      9.397854    2.094168    0.000000  
sample_2     10.090470    1.683023    0.000000  
sample_3      9.684365    3.292001    0.000000  
sample_4      9.461191    5.110372    0.000000  
...                ...         ...         ...  
sample_796   10.092770    8.819269    0.000000  
sample_797    8.745983    9.659081    0.000000  
sample_798    9.466711    4.677458    0.586693  
sample_799   10.400581    5.718751    0.000000  
sample_800    9.844794    4.550716    0.000000  

[801 rows x 20532 columns]

Ce code est utilisé pour trouver la moyenne des classes

In [561]:
# Groupez les données par classe
grouped = merged_df.groupby('Class')

# Calculez la moyenne pour chaque colonne
centre_de_toutes_les_classes = grouped.mean()

centre_de_toutes_les_classes

gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \
Class                                                                        
BRCA   0.011362  2.839739  2.805759  6.692613  10.003743     0.0  7.503838   
COAD   0.022212  3.438381  3.238618  7.830324   9.933747     0.0  7.292393   
KIRC   0.046544  2.398129  2.447997  6.340925   9.446830     0.0  7.008021   
LUAD   0.041088  3.358260  3.457701  6.625187   9.781486     0.0  7.283121   
PRAD   0.026544  3.441041  3.971262  6.662433   9.752363     0.0  7.807088   

         gene_7    gene_8    gene_9  ...  gene_20521  gene_20522  gene_20523  \
Class                                ...                                       
BRCA   0.499902  0.021714  0.031654  ...    5.891983    9.004483   10.090170   
COAD   0.509116  0.000000  0.000000  ...    5.473999    8.679328    9.966375   
KIRC   0.645374  0.024824  0.000000  ...    6.431149    8.573289   10.037905   
LUAD   0.393179  0.015985  0.000000  ...    5.557246    8.536567    9.966566   
PRAD   0.448975  0.007497  0.009264  ...    5.926976    8.733751   10.145661   

       gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \
Class                                                                           
BRCA     6.214461    9.728951   11.748519   10.135163    9.774321    6.206215   
COAD     0.695664    9.121327   12.215350   10.294870    9.352634    3.616018   
KIRC     3.324251   10.072278   11.805309   10.204060    9.431097    5.099268   
LUAD     4.767049    9.675499   11.788373    9.962015    9.394807    5.277331   
PRAD     5.933344    9.841068   11.341440   10.267547    9.696780    5.849697   

       gene_20530  
Class              
BRCA     0.073071  
COAD     0.039924  
KIRC     0.133671  
LUAD     0.191586  
PRAD     0.035729  

[5 rows x 20531 columns]

Cette fonction sert à caculer la distance entre deux dataframes

In [562]:
def df_class(classe):
    df = merged_df[merged_df['Class'] == classe] # Sélectionnez les données de la classe choisie
    df = df.iloc[:, 1:] # On exclut la colonne 'Class'
    return df

df_class('BRCA').head(10)

gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \
sample_4      0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   
sample_8      0.0  3.992125  2.772730  6.546692  10.488252     0.0  7.690222   
sample_10     0.0  3.492071  3.553373  7.151707  10.253446     0.0  8.301258   
sample_13     0.0  1.551048  3.529846  6.326825  10.633849     0.0  8.944659   
sample_14     0.0  1.964842  2.183010  6.596832  10.248141     0.0  7.087251   
sample_15     0.0  2.901379  3.685368  6.669665   9.999098     0.0  6.948834   
sample_20     0.0  2.066916  2.619953  6.757997  11.123940     0.0  5.415674   
sample_23     0.0  3.121844  2.473943  7.119616  10.732490     0.0  5.506031   
sample_27     0.0  0.657091  1.026304  5.411196   9.601388     0.0  6.263874   
sample_28     0.0  3.670081  3.382792  7.581939   9.386658     0.0  7.140523   

             gene_7  gene_8    gene_9  ...  gene_20521  gene_20522  \
sample_4   0.360982     0.0  0.000000  ...    5.996032    8.891425   
sample_8   0.352307     0.0  4.067604  ...    6.721974    9.597533   
sample_10  0.000000     0.0  0.000000  ...    6.231006    9.061674   
sample_13  0.000000     0.0  0.000000  ...    3.920732    7.846901   
sample_14  0.441483     0.0  0.000000  ...    6.078847   10.829588   
sample_15  0.000000     0.0  0.000000  ...    6.348744    9.234868   
sample_20  0.000000     0.0  0.000000  ...    5.475669    9.852696   
sample_23  0.000000     0.0  0.000000  ...    4.029073    7.571919   
sample_27  0.000000     0.0  0.000000  ...    6.742842    9.272040   
sample_28  0.000000     0.0  0.000000  ...    4.807221    8.195884   

           gene_20523  gene_20524  gene_20525  gene_20526  gene_20527  \
sample_4    10.373790    7.181162    9.846910   11.922439    9.217749   
sample_8     9.763753    7.933278   10.952880   12.498919   10.389954   
sample_10   10.042261    6.580103    9.785004   12.383834    9.869359   
sample_13    9.379835    7.195800    8.907645   12.079481    9.218936   
sample_14   10.404567    6.588536    8.952596   10.461725   10.582857   
sample_15    9.983891    5.252472    9.687888   12.068835    9.973008   
sample_20    9.783769    5.846676   10.806791   11.009122    8.984886   
sample_23    9.786257    5.666995    9.416534   11.917820    9.382985   
sample_27    9.774850    4.723340    9.922495   11.503865    9.485191   
sample_28   10.024447    7.921948    9.414692   11.762760    9.497850   

           gene_20528  gene_20529  gene_20530  
sample_4     9.461191    5.110372    0.000000  
sample_8    10.390255    7.828321    0.000000  
sample_10    8.991768    5.735121    0.892818  
sample_13    8.104101    3.660894    0.000000  
sample_14    9.488141    3.261967    0.000000  
sample_15    9.836426    9.052470    0.000000  
sample_20   10.356870    5.772084    0.000000  
sample_23    8.525258    3.686769    0.000000  
sample_27    9.435624    4.228888    0.000000  
sample_28   10.466668    4.028825    0.448690  

[10 rows x 20531 columns]

In [563]:
def inv_matrice_covariance(dataframe):
    matrice_cov = dataframe.cov()
    inv_matrice = np.linalg.inv(matrice_cov)
    return inv_matrice

In [564]:
print(merged_df.shape)

(801, 20532)


On essayant de lancer cette fonction on se rend compte que c'est impossible de prendre en compte toutes les gènes. Le temps de calcul devient irraisonnalbe donc il faut choisir les gènes les plus interessant afin de réduire considérablement le temps de calcul

In [565]:
#matrice_covariance(df_class('BRCA'))

In [566]:
def compter_colonnes_constantes(dataframe):
    colonnes_constantes = 0
    for colonne in dataframe.columns:
        if dataframe[colonne].nunique() == 1:
            colonnes_constantes += 1
    return colonnes_constantes

In [567]:
print(compter_colonnes_constantes(merged_df))
print(merged_df.shape)

267
(801, 20532)


Comme première idée on pense à supprimer les colonnes avec des valeurs constantes mais on se rend compte que leurs nombre est très petit devant le nombre de colonnes totales 267<<20532. Donc on se retourne vers une autre méthode

In [568]:
def variance(dataframe):
    var = dataframe.var()
    return var

In [569]:
def variance_significatifs(dataframe,seuil):
    variances = variance(dataframe)
    variances_significatifs = variances[variances > seuil]
    return variances_significatifs

In [570]:
def dataframe_sans_variances_insignificatifs(dataframe, variances_significatifs):
    # Étape 1 : Convertir la première DataFrame en une liste de noms de gènes
    genes_a_garder = list(variances_significatifs.index)
    # Étape 2 : Sélectionner uniquement les colonnes qui ne sont pas dans la liste de gènes de la première DataFrame
    df_sans_genes_insignificatifs = dataframe[genes_a_garder]
    return df_sans_genes_insignificatifs

In [571]:
def dataframe_variances_significatifs(dataframe, seuil):
    
    variances_significatifs = variance_significatifs(dataframe, seuil)
    df_sans_genes_df1 = dataframe_sans_variances_insignificatifs(dataframe, variances_significatifs)

    return df_sans_genes_df1

In [572]:
df_significant = dataframe_variances_significatifs(merged_df, 6)
inv_matrice_covariance(df_significant)

/tmp/ipykernel_116161/1253365460.py:2: FutureWarning: The default value of numeric_only in DataFrame.var is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  var = dataframe.var()


array([[ 3.45111316e+12,  1.81354711e+13, -4.28690996e+11, ...,
        -7.37747990e+12, -2.62012588e+13, -4.04380403e+12],
       [-1.42334574e+13,  1.22011907e+13,  6.88086012e+12, ...,
         8.65932385e+12, -1.59491087e+13, -1.72220935e+13],
       [-9.44506357e+12,  5.88928252e+12, -6.72174159e+12, ...,
         1.91538002e+13,  4.39486925e+12, -1.12192759e+13],
       ...,
       [ 7.26361232e+12, -2.47161037e+12, -6.32440658e+12, ...,
         7.54202621e+12, -4.44265854e+12,  1.15647859e+12],
       [ 8.87113500e+12, -4.62320569e+13,  7.89025751e+12, ...,
        -3.88963578e+13,  2.25203722e+13,  3.31146194e+13],
       [ 2.26291620e+12, -1.48436205e+13,  3.21112764e+12, ...,
        -6.67286091e+12,  4.96535230e+12,  5.30819704e+12]])

In [654]:
def distance(df1, df2, type_distance='euclidienne'):
    if type_distance == 'euclidienne':
        df_distance = abs(df1**2 - df2**2)
        return np.sqrt(df_distance.transpose().sum())
    
    if type_distance == 'cosinus':
        similarité_cosinus = np.dot(df1, df2) / (np.linalg.norm(df1) * np.linalg.norm(df2))
        return 1-similarité_cosinus
    
    if type_distance == 'mahalanobis':
        variances_significatifs = variance_significatifs(df1, 10)
        df1_sans_genes_insignificatifs = dataframe_sans_variances_insignificatifs(df1, variances_significatifs)
        df2_sans_genes_insignificatifs = dataframe_sans_variances_insignificatifs(df2, variances_significatifs)
        matrice_de_covariance = inv_matrice_covariance(df1_sans_genes_insignificatifs)
        # Créer une liste pour stocker les résultats de chaque calcul
        resultats = []

        # Boucler à travers chaque ligne de df1_sans_genes_insignificatifs
        for index, ligne in df1_sans_genes_insignificatifs.iterrows():
            # Effectuer l'opération pour cette ligne
            step1_result = ligne.values.dot(matrice_de_covariance)
            final_result = step1_result.dot(df2_sans_genes_insignificatifs.transpose())
            # Ajouter le résultat à la liste des résultats
            resultats.append(final_result)

        # Créer une nouvelle DataFrame à partir des résultats
        nouvelle_dataframe = pd.DataFrame(resultats)
            
        # Assigner les mêmes index que ceux de df1_sans_genes_insignificatifs
        nouvelle_dataframe.index = df1_sans_genes_insignificatifs.index
        
        return nouvelle_dataframe

In [655]:
def distance_intra(Class, type_de_distance='euclidienne'):
    
    if Class not in ['BRCA', 'KIRC', 'LUAD', 'PRAD']:
        return "Class not found"
    if type_de_distance not in ['euclidienne', 'mahalanobis', 'cosinus']:
        return "Type de distance non reconnu"
    
    df_classe = df_class(Class)
    
    centre_de_la_classe = centre_de_toutes_les_classes.loc[Class]
    
    # Calculez la distance intra-classe
    df_distance = distance(df_classe, centre_de_la_classe, type_de_distance) # cette dataframe contient toutes les différences entre tous les patients de la classe et le centre de la classe
    distance_max_entre_le_centre_et_un_patient = float(df_distance.max())
    
    return distance_max_entre_le_centre_et_un_patient

In [656]:
distance_intra('LUAD')

543.5869315923811